### PR TITLE
Integrate mailbox into runtime firmware

### PR DESCRIPTION
**Pull request by jhand2**
Handle mailbox commands in runtime and add test which can run against
sw and verilator.

### CODE REVIEW

File: runtime/tests/integration_tests.rs [hunk 1]
```diff
@@ -31,6 +32,35 @@ fn run_rt_test(test_bin_name: Option<&str>) {
...
+    model.soc_mbox().cmd().write(|_| cmd_id);
+    model.soc_mbox().dlen().write(|_| arg_buf.len().try_into().unwrap());
+
+    for word in arg_buf {
+        model.soc_mbox().datain().write(|_| *word);
+    }
+    model.soc_mbox().execute().write(|exec_reg| exec_reg.execute(true));
+
+    model.step_until(|m| {
+        m.soc_mbox().status().read().status().cmd_complete()
```
Comment (korran) [inline runtime/tests/integration_tests.rs:50]: Should be:
model.step_until(|m| {
        !m.soc_mbox().status().read().status().cmd_busy()).

File: runtime/src/commands.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+pub fn handle_mailbox_commands() -> CaliptraResult<()> {
+    loop {
+        wait_for_cmd();
+
+        if let Some(mut recv_txn) = Mailbox::default().try_start_recv_txn() {
+            let cmd_id = recv_txn.cmd();
+            let len = recv_txn.dlen() as usize;
+            let mut buf = [0u32; 1024];
+            recv_txn.copy_request(0, &mut buf[..len]).unwrap();
+            recv_txn.complete(MboxStatusE::CmdBusy);
```
Comment (korran) [inline runtime/src/commands.rs:23]: This is wrong; mailbox response needs to be populated before calling complete.

The driver doesn't appear to have an API for that

File: runtime/src/commands.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+            // TODO: Actually handle command
+            cprintln!("[rt] Received command={}, len={}", cmd_id, len);
+
+            // Write response
+            let out_buf = [0xFFu8; 16];
+            let mut send_txn = Mailbox::default().try_start_send_txn().unwrap();
```
Comment (korran) [inline runtime/src/commands.rs:30]: You don't want to use try_start_send_txn to send a response; that's used for making an "RPC" request to the SoC.

File: drivers/src/mailbox.rs [hunk 4]
```diff
@@ -70,10 +69,44 @@ impl Mailbox {
     /// * 'MailboxRecvTxn' - Object representing a receive operation
     pub fn try_start_recv_txn(&self) -> Option<MailboxRecvTxn> {
         let mbox = mbox::RegisterBlock::mbox_csr();
-        match mbox.status().read().mbox_fsm_ps() {
-            MboxFsmE::MboxExecuteUc => Some(MailboxRecvTxn::default()),
-            _ => None,
+        if mbox.status().read().mbox_fsm_ps().mbox_execute_uc() {
+            match mbox.status().read().status() {
```
Comment (korran) [inline drivers/src/mailbox.rs:73]: What is the purpose of this change?

File: drivers/src/mailbox.rs [hunk 5]
```diff
@@ -70,10 +69,44 @@ impl Mailbox {
...
+        } else {
+            None
+        }
+    }
+}
+
+pub struct MailboxTxn;
+
+impl MailboxTxn {
+    pub(crate) fn enqueue(buf: &[u8]) -> CaliptraResult<()> {
```
Comment (korran) [inline drivers/src/mailbox.rs:86]: This should probably just be a top-level function.

File: drivers/src/mailbox.rs (hunk 4 above)
Comment (jhand2) [reply drivers/src/mailbox.rs:73]: In the loop in `runtime/src/main.rs`, I'm continually calling `try_start_recv_txn` under the assumption that if there's a command in the mailbox and we're in execute state, we've received a new command.

The problem is that when we write data back to the mailbox, it's still in execute state. So until the SoC clears the execute register, runtime fw will keep looping and trying to execute the same command repeatedly.

I'm open to a different way of doing this though.

File: drivers/src/mailbox.rs (hunk 5 above)
Comment (jhand2) [reply drivers/src/mailbox.rs:86]: Ack. Will do.

File: drivers/src/mailbox.rs (hunk 4 above)
Comment (korran) [reply drivers/src/mailbox.rs:73]: In the RTL, once you write to mbox.status(), mbox.mbox_fsm_ps() switches from ExecuteUc to ExecuteSoc. Apparently this is not happening in the sw-emulator.

Comment (jhand2) [reply drivers/src/mailbox.rs]: I removed this. I still think the emulator needs to be updated, but that's not blocking this PR (mostly because the test steps until the status is set and then stops).

I'm happy to make this change in this PR or just do it in a follow-up.

File: drivers/src/mailbox.rs (hunk 5 above)
Comment (jhand2) [reply drivers/src/mailbox.rs]: Moved runtime mailbox usage outside of this driver

File: drivers/src/mailbox.rs (hunk 4 above)
Comment (jhand2) [reply drivers/src/mailbox.rs]: Resolving for now

File: drivers/src/error.rs [hunk 6]
```diff
@@ -53,6 +53,9 @@ pub enum CaliptraComponent {
 
     /// CSRNG
     Csrng = 13,
+
+    /// Runtime firmware
+    Runtime = 14,
```
Comment (korran) [inline drivers/src/error.rs:58]: I don't think this is where @vsonims intended the runtime errors to go. Maybe put a TODO here saying something like "move this somewhere more idiomatic".

@rusty1968 has plans to revamp how the errors are organized.

Comment (jhand2) [reply drivers/src/error.rs:60]: Done (once https://github.com/chipsalliance/caliptra-sw/pull/220 is merged I'll switch to use that mechanism)

File: rom/dev/tests/test_idevid_derivation.rs [hunk 7]
```diff
@@ -55,7 +58,7 @@ fn test_idev_subj_key_id_algo() {
         hw.upload_firmware(&image_bundle.to_bytes().unwrap())
             .unwrap();
 
-        let result = hw.copy_output_until_exit_success(&mut output);
-        assert!(result.is_ok());
+        // Step until runtime firmware denotes successful firmware upload
+        hw.step_until(|m| m.soc_mbox().status().read().status().cmd_complete());
```
Comment (korran) [inline rom/dev/tests/test_idevid_derivation.rs:62]: The ROM still has plenty of stuff to do after the mailbox transaction is completed. We should probably be waiting until the runtime firmware has printed something out to the UART, or perhaps just send the runtime firmware a mailbox command to make sure it is alive.

File: runtime/tests/integration_tests.rs [hunk 8]
```diff
@@ -30,20 +47,84 @@ fn run_rt_test(test_bin_name: Option<&str>) {
...
+    let image = caliptra_builder::build_and_sign_image(
+        &FMC_WITH_UART,
+        &runtime_fwid,
+        ImageOptions::default(),
+    )
+    .unwrap();
+
+    model.step_until(|m| m.soc_ifc().cptra_flow_status().read().ready_for_fw());
+
+    model.upload_firmware(&image.to_bytes().unwrap()).unwrap();
```
Comment (korran) [inline runtime/tests/integration_tests.rs:59]: nitpick: you can get this part for free if you pass set BootParams::fw_image to Some(&image.to_bytes().unwrap()) above. See https://github.com/chipsalliance/caliptra-sw/blob/main/hw-model/src/lib.rs#L185

File: test/tests/smoke_test.rs [hunk 9]
```diff
@@ -30,7 +31,13 @@ fn smoke_test() {
     })
     .unwrap();
     let mut output = vec![];
-    hw.copy_output_until_exit_success(&mut output).unwrap();
+
+    // Step until runtime firmware denotes successful firmware upload
+    hw.step_until(|m| m.soc_mbox().status().read().status().cmd_complete());
```
Comment (korran) [inline test/tests/smoke_test.rs:36]: Does this actually do what you think it does? hw.upload_firmware() already waits for the upload command to complete:

https://github.com/chipsalliance/caliptra-sw/blob/effcd075176fdb598ba103a43e7aa99ac9f06b09/hw-model/src/lib.rs#L385-L399

I don't think the strings we are looking for below will be found.

File: runtime/src/main.rs [hunk 10]
```diff
@@ -31,11 +32,21 @@ const BANNER: &str = r#"
                |_|
 "#;
 
+fn report_fw_load_success() {
+    // Complete the firmware download mailbox command
+    let mbox = mbox::RegisterBlock::mbox_csr();
+    mbox.status()
+        .write(|w| w.status(|_| MboxStatusE::CmdComplete));
```
Comment (korran) [inline runtime/src/main.rs:39]: This has already been done. The ROM completed the mailbox transaction write after it validated the signature. Remove this.

File: rom/dev/tests/test_idevid_derivation.rs (hunk 7 above)
Comment (jhand2) [reply rom/dev/tests/test_idevid_derivation.rs]: Done.

File: test/tests/smoke_test.rs (hunk 9 above)
Comment (jhand2) [reply test/tests/smoke_test.rs]: Discussed offline. Switched to checking for UART output from RT.

File: runtime/src/main.rs (hunk 10 above)
Comment (jhand2) [reply runtime/src/main.rs]: Done.

File: runtime/tests/integration_tests.rs (hunk 8 above)
Comment (jhand2) [reply runtime/tests/integration_tests.rs]: Done.

**Review by korran — APPROVED** [review]
